### PR TITLE
Teapot support

### LINKED ISSUE #428: Statuscat 418 doesn't work
## Description
Http code 418 is reported to be invalid, even though its defined [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418). A [cat](https://http.cat/418.jpg) exists for the status code too.

## Steps to Reproduce
Send `.statuscat 418` to Seasonalbot.

## Expected Behaviour
Sends an embed with [this](https://http.cat/418.jpg) image in it, and a header of `Status: 418`.

## Actual Behaviour
Code not found:
image

## Known Impacted Platforms

- [x] Web
- [x] Desktop
- [x] Android App
- [x] iOS App

## Possible Solutions
Check for status code 418 before checking for the other codes the http library.

## Would you like to implement a fix?

***Note: For high-priority or critical bugs, fixes may be implemented by staff.***

- [x] I'd like to implement the bug fix
- [ ] Anyone can implement the bug fix

### LINKED ISSUE #500: httpcat missing codes
## Description

The `.http_cat` command is supposed to send the picture associated with the status code the user enters. To check if the user has entered a valid status code, it checks if it is in the [`HTTPStatus` enum](https://docs.python.org/3/library/http.html#http.HTTPStatus).  The problem with this, is that three of the status codes, `103 EARLY_HINTS`, `418 IM_A_TEAPOT` and `425 TOO_EARLY`, are new to python version 3.9, though they are in [http.cat](https://http.cat/). We should fix it so that these codes can be used.

## Reasoning

More codes for the `.http_cat` command.

## Proposed Implementation

There are two ways I can think of implementing this. 

- Create a list of valid codes from  [http.cat](https://http.cat/).
- Or make a new class deriving from [`HTTPStatus`](https://docs.python.org/3/library/http.html#http.HTTPStatus) with the extra codes.

I think the first way would be easier to implement and maybe even slightly more performant.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [ ] Anyone can implement this feature

### LINKED ISSUE #608: HTTP Status Code 418 not recognized by .http_status command
## Description
The .http_status .status and .httpstatus commands do not recognize HTTP status code 418 (I am a Teapot) as described in [RFC-2324 : Hyper Text Coffee Pot Control Protocol (HTCPCP/1.0)
](https://tools.ietf.org/html/rfc2324)

> 2.3.2 418 I'm a teapot
>   Any attempt to brew coffee with a teapot should result in the error
>   code "418 I'm a teapot". The resulting entity body MAY be short and
>   stout.

## Steps to Reproduce
Input any of the following commands in the sir-lancebot-playground channel
```
.httpstatus dog 418
.httpstatus cat 418

.status cat 418
.status dog 418

.http_status cat 418
.http_status dog 418
```

## Expected Behaviour
The relevant image from either https://http.cat/418 
http-cat
or https://httpstatusdogs.com/418-im-a-teapot

418-dogs
Should be displayed

## Actual Behaviour
An error message is displayed indicating that the status code 418 does not exist:
Screen Shot 2021-03-02 at 3 10 55 PM

## Known Impacted Platforms

- [X] Web
- [X] Desktop
- [x] Android App
- [x] iOS App

## Possible Solutions

Support for HTTP Status Code 418 Is currently only available in [Python 3.9](https://docs.python.org/3.9/library/http.html). 

Screen Shot 2021-03-02 at 3 29 43 PM

Screen Shot 2021-03-02 at 3 30 04 PM

 Implement manual workaround for this status code until default interpreter is upgraded to 3.9
## Additional Details

Relevant Python Bug : https://bugs.python.org/issue39507

## Would you like to implement a fix?

***Note: For high-priority or critical bugs, fixes may be implemented by staff.***

- [X] I'd like to implement the bug fix
- [x] Anyone can implement the bug fix

### PR DESCRIPTION
**Pull request by sam-heller**
## Relevant Issues
Closes #428 
Closes #500
Closes #608

## Description
Removed HTTPStatus from the equation entirely in favor of just relying on the cat/dog servers to send us back errors if we get bad input. 

Minor refactor to pull strings out of the executable code into variables at the top. 

## Reasoning
Implemented design and strategy suggested by staff in #608 discussion. 

## Screenshots
Screen Shot 2021-03-03 at 8 16 13 AM
Screen Shot 2021-03-03 at 8 16 18 AM

## Did you:

- [X] Join the [**Python Discord Community**](https://discord.gg/python)?
- [X] If dependencies have been added or updated, run `pipenv lock`?
- [X] **Lint your code** (`pipenv run lint`)?
- [ ] Set the PR to **allow edits from contributors**? *Edits allowed from maintainers, contributors doesn't appear to be an option*

### CODE REVIEW

File: bot/exts/evergreen/status_codes.py [hunk 1]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
-
-        except ValueError:
-            embed.set_footer(text='Inputted status code does not exist.')
-
-        except NotImplementedError:
-            embed.set_footer(text='Inputted status code is not implemented by http.cat yet.')
-
-        finally:
-            await ctx.send(embed=embed)
+        """Assemble Cat URL and build Embed."""
```
Comment (jb3) [inline bot/exts/evergreen/status_codes.py:25]: This docstring is used in the `.help` documentation, so we should keep it more human friendly.
```suggestion
        """Send a cat version of the requested HTTP status code."""
```

File: bot/exts/evergreen/status_codes.py [hunk 2]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
-
-        except ValueError:
-            embed.set_footer(text='Inputted status code does not exist.')
-
-        except NotImplementedError:
-            embed.set_footer(text='Inputted status code is not implemented by httpstatusdogs.com yet.')
-
-        finally:
-            await ctx.send(embed=embed)
+        """Assemble Dog URL and build Embed."""
```
Comment (jb3) [inline bot/exts/evergreen/status_codes.py:30]: Same as above
```suggestion
        """Send a dog version of the requested HTTP status code."""
```

File: bot/exts/evergreen/status_codes.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
-from http import HTTPStatus
-
 import discord
 from discord.ext import commands
 
 HTTP_DOG_URL = "https://httpstatusdogs.com/img/{code}.jpg"
 HTTP_CAT_URL = "https://http.cat/{code}.jpg"
+STATUS_TEMPLATE = '**Status: {code}**'
+ERR_404 = 'Unable to find status Floof for {code}.'
+ERR_UNKNOWN = 'Error attempting to retrieve status Floof for {code}.'
```
Comment (jb3) [inline bot/exts/evergreen/status_codes.py:8]:
```suggestion
ERR_404 = 'Unable to find status floof for {code}.'
ERR_UNKNOWN = 'Error attempting to retrieve status floof for {code}.'
```

**Review by jb3 — CHANGES_REQUESTED** [review]
:cat: The most important PR ever received :tea:

Some minor comments on wording, tested locally and all works well.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
### Bug 🙃 

Screenshot from 2021-03-03 19-38-45

**Comment by ChrisLovering** [comment]
> ### Bug 🙃
> Screenshot from 2021-03-03 19-38-45

Could you list some reproduction steps to help @sam-heller with fixing it?

**Comment by ChrisLovering** [comment]
Also, I'm not sure about passing user input directly into a web request like this, without sanitising it first.
I also realise this is the way is was before this PR too, but looks like something we should fix.

**Comment by Shivansh-007** [comment]
> @Shivansh-007 Could you list some reproduction steps to help @sam-heller with fixing it?

We can make use `textwrap` module and then do something like this:
```py
textwrap.shorten(code width=50, placeholder="...")
```

**Comment by jb3** [comment]
We could also just validate that codes are below some sane value, like 999.

Convert to an integer and handle any failure there, then check the integer is greater than 0 and less than 999, we don’t need to bring textwrap into it.

File: bot/exts/evergreen/status_codes.py [hunk 4]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
+            if 200 <= response.status <= 299:
+                await ctx.send(embed=discord.Embed(
+                    title=STATUS_TEMPLATE.format(code=code)).set_image(url=url))
+            elif 404 == response.status:
+                await ctx.send(embed=discord.Embed(
+                    title=ERR_404.format(code=code)).set_image(url=url)
+                )
+            else:
+                await ctx.send(embed=discord.Embed(
+                    title=STATUS_TEMPLATE.format(code=code), footer=ERR_UNKNOWN.format(code=code))
```
Comment (Numerlor) [inline bot/exts/evergreen/status_codes.py:45]: This doesn't work, you'll have to set the footer through `embed_obj.set_footer(text=...)`

File: bot/exts/evergreen/status_codes.py [hunk 5]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
+    async def build_embed(self, url: str, ctx: commands.Context, code: int) -> None:
+        """Attempt to build and dispatch embed. Append error message instead if something goes wrong."""
+        async with self.bot.http_session.get(url, allow_redirects=False) as response:
+            if 200 <= response.status <= 299:
+                await ctx.send(embed=discord.Embed(
+                    title=STATUS_TEMPLATE.format(code=code)).set_image(url=url))
+            elif 404 == response.status:
+                await ctx.send(embed=discord.Embed(
+                    title=ERR_404.format(code=code)).set_image(url=url)
+                )
```
Comment (Numerlor) [inline bot/exts/evergreen/status_codes.py:42]: This won't catch invalid dog codes, as those redirect to their 404 image page through the 302 response code, and will go to the generic handler

**Review by Numerlor — CHANGES_REQUESTED** [review]
The style for hanging indents in the projects is to have the closing parentheses on their own line which isn't used in the `build_embed` if chain. I believe only creating the embed objects in the if bodies and then having a send at the end of the function will look better and won't need any splitting up of the lines

File: bot/exts/evergreen/status_codes.py [hunk 6]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
+        """Assemble Dog URL and build Embed."""
+        await self.build_embed(url=HTTP_DOG_URL.format(code=code), ctx=ctx, code=code)
+
+    async def build_embed(self, url: str, ctx: commands.Context, code: int) -> None:
+        """Attempt to build and dispatch embed. Append error message instead if something goes wrong."""
+        async with self.bot.http_session.get(url, allow_redirects=False) as response:
+            if 200 <= response.status <= 299:
+                await ctx.send(embed=discord.Embed(
+                    title=STATUS_TEMPLATE.format(code=code)).set_image(url=url))
+            elif 404 == response.status:
```
Comment (Numerlor) [inline bot/exts/evergreen/status_codes.py:39]: The yoda condition looks a bit weird to me as I don't think I've seen it anywhere else in the codebase but I suppose it's fine

Comment (sam-heller) [reply bot/exts/evergreen/status_codes.py:39]: Taking another look and it's definitely not the most readable. What do you think about swapping it with
```python
if response.status in range(200,299):
```
?

Comment (ToxicKidz) [reply bot/exts/evergreen/status_codes.py:39]: [`ClientResponse.ok`](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientResponse.ok) exists.

File: bot/exts/evergreen/status_codes.py (hunk 3 above)
Comment (ToxicKidz) [inline bot/exts/evergreen/status_codes.py:8]: These should be double quotes too
```suggestion
STATUS_TEMPLATE = "**Status: {code}**"
ERR_404 = "Unable to find status Floof for {code}."
ERR_UNKNOWN = "Error attempting to retrieve status Floof for {code}."
```

File: bot/exts/evergreen/status_codes.py (hunk 5 above)
Comment (ToxicKidz) [inline bot/exts/evergreen/status_codes.py:42]: The `embed=` part should be on a new line.
```suggestion
                await ctx.send(
                    embed=discord.Embed(
                        title=ERR_404.format(code=code)
                    ).set_image(url=url)
                )
```

File: bot/exts/evergreen/status_codes.py [hunk 7]
```diff
@@ -21,49 +22,28 @@ async def http_status_group(self, ctx: commands.Context) -> None:
...
+                await ctx.send(embed=discord.Embed(
+                    title=ERR_404.format(code=code)).set_image(url=url)
+                )
+            else:
+                await ctx.send(embed=discord.Embed(
+                    title=STATUS_TEMPLATE.format(code=code), footer=ERR_UNKNOWN.format(code=code))
+                )
```
Comment (ToxicKidz) [inline bot/exts/evergreen/status_codes.py:46]: The `embed=` part should be on a new line, and `footer` isn't a valid keyword argument for `Embed`. Use `Embed.set_footer` instead.
```suggestion
                await ctx.send(
                    embed=discord.Embed(
                        title=STATUS_TEMPLATE.format(code=code)
                    ).set_footer(text=ERR_UNKNOWN.format(code=code))
                )
```

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

**Comment by Xithrius** [comment]
@sam-heller will you be able to resolve these requests for changes?

**Comment by Xithrius** [comment]
This user has not been active for some time on GitHub. This PR is now up for grabs, just ping me if you'd like to take it.

**Comment by brad90four** [comment]
Hey @Xithrius , I can take a stab at this I think. Looks like all of the changes requested are pretty clear. Not sure what you mean by file conflicts and reviews that need to be resolved but I could probably figure it out with a little bit of help.

**Comment by Xithrius** [comment]
@brad90four Hello! Sorry for the late reply. I'll assign this to you. In the meantime if you need help with conflicting files, I'd go into the #dev-contrib channel so people are able to get the most amount of information into your brain about how to deal with them. You can see conflicting files at near the end of this PR. In this case, it should show up as such:

image

When someone reviewed a PR, it should look like the screenshot below. Once you see it in this PR, you can start going over what the user has suggested/requested/commented. You can commit multiple items in batches when in the "Files changed" tab, or just commit directly with the commit button that is shown next to a suggestion.

image

**Pull request closed without merging**